src/Microsoft.Health.Fhir.SqlServer/Features/Search/HashingSqlQueryParameterManager.cs

```diff
@@ -0,0 +1,201 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using EnsureThat;
+using Microsoft.Data.SqlClient;
+using Microsoft.Health.SqlServer;
+using Microsoft.Health.SqlServer.Features.Schema.Model;
+using Microsoft.Health.SqlServer.Features.Storage;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search
+{
+    /// <summary>
+    /// Wraps a <see cref="SqlQueryParameterManager"/>, adding the ability to compute a hash of a subset of the parameters.
+    /// </summary>
+    public class HashingSqlQueryParameterManager
+    {
+        private readonly SqlQueryParameterManager _inner;
+        private readonly HashSet<SqlParameter> _setToHash = new();
+
+        public HashingSqlQueryParameterManager(SqlQueryParameterManager inner)
+        {
+            EnsureArg.IsNotNull(inner, nameof(inner));
+            _inner = inner;
+        }
+
+        public bool HasParametersToHash => _setToHash.Count > 0;
+
+        public SqlParameter AddParameter<T>(Column<T> column, T value, bool includeInHash = true)
+        {
+            return AddParameter((Column)column, value, includeInHash);
+        }
+
+        public SqlParameter AddParameter(Column column, object value, bool includeInHash = true)
+        {
+            SqlParameter parameter = _inner.AddParameter(column, value);
+            if (includeInHash)
+            {
+                _setToHash.Add(parameter);
+            }
+
+            return parameter;
+        }
+
+        public SqlParameter AddParameter(object value, bool includeInHash = true)
+        {
+            SqlParameter parameter = _inner.AddParameter(value);
+            if (includeInHash)
+            {
+                _setToHash.Add(parameter);
+            }
+
+            return parameter;
+        }
+
+        /// <summary>
+        /// Appends a Base64-encoded SHA-256 hash of the parameters currently added to this instance with includeInHash = true
+        /// </summary>
+        /// <param name="stringBuilder">A string builder to append the hash to.</param>
+        public void AppendHash(IndentedStringBuilder stringBuilder)
+        {
+            IncrementalHash incrementalHash = null;
+            Span<byte> buf = stackalloc byte[256];
+            int currentBufferIndex = 0;
+
+            foreach (SqlParameter sqlParameter in _setToHash)
+            {
+                if (!_setToHash.Contains(sqlParameter))
```

Comment:
>if (!_setToHash.Contains(sqlParameter)) [](start = 16, length = 39)

Is there a reason we have this check here since we looping through `_setToHash` ?

Reply:
Ha! yes, good point